helix-core/src/search.rs

```diff
@@ -61,3 +61,184 @@ pub fn find_nth_prev(
+    right_rdfa: DenseDFA<Vec<usize>, usize>,
+
+    /// Locate start of match searching left.
+    left_fdfa: DenseDFA<Vec<usize>, usize>,
+    /// Locate end of match searching left.
+    left_rdfa: DenseDFA<Vec<usize>, usize>,
+}
+
+impl Searcher {
+    pub fn new(pattern: &str) -> Result<Searcher, RegexError> {
+        // Check case info for smart case
+        let has_uppercase = pattern.chars().any(|c| c.is_uppercase());
+
+        // Create Regex DFAs for all search directions.
+        let mut builder = dense::Builder::new();
+        let builder = builder.case_insensitive(!has_uppercase);
+
+        let left_fdfa = builder.clone().reverse(true).build(pattern)?;
+        let left_rdfa = builder
+            .clone()
+            .anchored(true)
+            .longest_match(true)
+            .build(pattern)?;
+
+        let right_fdfa = builder.clone().build(pattern)?;
+        let right_rdfa = builder
+            .anchored(true)
+            .longest_match(true)
+            .reverse(true)
+            .build(pattern)?;
+
+        Ok(Searcher {
+            right_fdfa,
+            right_rdfa,
+            left_fdfa,
+            left_rdfa,
+        })
+    }
+    pub fn search_prev(&self, text: RopeSlice, offset: usize) -> Option<Range<usize>> {
+        let text = text.slice(..offset);
+        let start = self.rfind(text, &self.left_fdfa)?;
+        let end = self.find(text.slice(start..), &self.left_rdfa)?;
+
+        Some(start..start + end)
+    }
+
+    pub fn search_next(&self, text: RopeSlice, offset: usize) -> Option<Range<usize>> {
+        let text = text.slice(offset..);
+        let end = self.find(text, &self.right_fdfa)?;
+        let start = self.rfind(text.slice(..end), &self.right_rdfa)?;
+
+        Some(offset + start..offset + end)
+    }
+
+    /// Returns the end offset of the longest match. If no match exists, then None is returned.
+    /// NOTE: based on DFA::find_at
+    fn find(&self, text: RopeSlice, dfa: &impl DFA) -> Option<usize> {
+        // TODO: check this inside main search
+        // if dfa.is_anchored() && start > 0 {
+        //     return None;
+        // }
+
+        let mut state = dfa.start_state();
+        let mut last_match = if dfa.is_dead_state(state) {
+            return None;
+        } else if dfa.is_match_state(state) {
+            Some(0)
+        } else {
+            None
+        };
+
+        let mut chunk_byte_offset = 0;
+
+        for chunk in text.chunks() {
+            for (i, &b) in chunk.as_bytes().iter().enumerate() {
+                state = unsafe { dfa.next_state_unchecked(state, b) };
+                if dfa.is_match_or_dead_state(state) {
+                    if dfa.is_dead_state(state) {
+                        return last_match;
+                    }
+                    last_match = Some(chunk_byte_offset + i + 1);
+                }
+            }
+            chunk_byte_offset += chunk.len();
+        }
+
+        last_match
```

Comment:
Not sure sure how good this part is. Did you took it from https://github.com/alacritty/alacritty/blob/3e867a056018c507d79396cb5c5b4b8309c609c2/alacritty_terminal/src/term/search.rs#L184-L282?

Reply:
It looks about right to me, although you should definitely write tests for it.

Also, I'd suggest using `dfa.next_state` and avoid the `unsafe` unless you can witness a measurable perf improvement. The only thing `next_state_unchecked` does is elide bound checks.

Reply:
@pickfire This is almost a direct copy of `DFA::find_at`/`rfind_at`, I've mainly just added another loop that feeds chunks into the dfa.